storage/memory/tree_storage.go

```diff
@@ -0,0 +1,278 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package memory
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+	"strings"
+	"sync"
+
+	"github.com/golang/glog"
+	"github.com/google/btree"
+	"github.com/google/trillian"
+	"github.com/google/trillian/storage"
+	"github.com/google/trillian/storage/cache"
+	"github.com/google/trillian/storage/storagepb"
+)
+
+const degree = 8
+
+func subtreeKey(treeID, rev int64, nodeID storage.NodeID) btree.Item {
+	return &kv{k: fmt.Sprintf("/%d/subtree/%s/%d", treeID, nodeID.String(), rev)}
+}
+
+// tree stores all data for a given treeID
+type tree struct {
+	mu    sync.RWMutex
+	store *btree.BTree
+	// currentSTH is the timestamp of the current STH.
+	currentSTH int64
+	meta       *trillian.Tree
+}
+
+func (t *tree) Lock() {
+	t.mu.Lock()
+}
+
+func (t *tree) Unlock() {
+	t.mu.Unlock()
+}
+
+func (t *tree) RLock() {
+	t.mu.RLock()
+}
+
+func (t *tree) RUnlock() {
+	t.mu.RUnlock()
+}
+
+// Dump ascends the tree, logging the items contained.
+func Dump(t *btree.BTree) {
+	t.Ascend(func(i btree.Item) bool {
+		glog.Infof("%#v", i)
+		return true
+	})
+}
+
+// memoryTreeStorage is shared between the memoryLog and (forthcoming) memoryMap-
+// Storage implementations, and contains functionality which is common to both,
+type memoryTreeStorage struct {
+	mu    sync.RWMutex
+	trees map[int64]*tree
+}
+
+func newTreeStorage() *memoryTreeStorage {
+	return &memoryTreeStorage{
+		trees: make(map[int64]*tree),
+	}
+}
+
+// getTree returns the tree associated with id, or nil if no such tree exists.
+func (m *memoryTreeStorage) getTree(id int64) *tree {
+	m.mu.RLock()
+	defer m.mu.RUnlock()
+	return m.trees[id]
+}
+
+// kv is a simple key->value type which implements btree's Item interface.
+type kv struct {
+	k string
+	v interface{}
+}
+
+// Less than by k's string key
+func (a kv) Less(b btree.Item) bool {
+	return strings.Compare(a.k, b.(*kv).k) < 0
+}
+
+// newTree creates and initializes a tree struct.
+func newTree(t trillian.Tree) *tree {
+	ret := &tree{
+		store: btree.New(degree),
+		meta:  &t,
+	}
+	k := unseqKey(t.TreeId)
+	k.(*kv).v = list.New()
+	ret.store.ReplaceOrInsert(k)
+
+	k = hashToSeqKey(t.TreeId)
+	k.(*kv).v = make(map[string][]int64)
+	ret.store.ReplaceOrInsert(k)
+
+	return ret
+}
+
+func (m *memoryTreeStorage) beginTreeTX(ctx context.Context, readonly bool, treeID int64, hashSizeBytes int, strataDepths []int, populate storage.PopulateSubtreeFunc, prepare storage.PrepareSubtreeWriteFunc) (treeTX, error) {
```

Comment:
Passed in a subtreeCache directly rather than all the bits necessary to build one [also from the mysql fork ;)]